File: Zig/day04/src/group.zig
pub fn Grouped(comptime T: type) type {
    return struct {
        slice: []T,

        const Self = @This();

        pub fn groups(slice: []T) Self {
            return Self{ .slice = slice };
        }

        pub fn next(self: *Self) ?[]T {
            if (self.slice.len == 0) {
                return null;
            }

            var fst = self.slice[0];

            var count = usize(0);

            for (self.slice) |elem| {
                if (!(elem == fst)) {
                    break;
                }
                count += 1;
            }

            var ret = self.slice[0..count];
            self.slice = self.slice[count..self.slice.len];
            return ret;
        }
    };
}

File: Rust/day04/src/group.rs
use std::mem;

pub struct Grouped<'a, T: 'a> {
    slice: &'a [T],
}

impl<'a, T: 'a> Grouped<'a, T> {
    pub fn groups(slice: &'a [T]) -> Self {
        Self { slice }
    }
}

impl<'a, T: Ord + 'a> Iterator for Grouped<'a, T> {
    type Item = &'a [T];

    fn next(&mut self) -> Option<Self::Item> {
        if self.slice.is_empty() {
            return None;
        }

        let mvd = mem::take(&mut self.slice);

        let (fst, res) = mvd.split_first()?;

        let count = res.iter().take_while(|elem| *elem == fst).count();

        let (ret, rest) = mvd.split_at(count + 1);

        self.slice = rest;

        Some(ret)
    }
}
